### DISCUSSION TITLE
Geolocation widget in Budibase

### DISCUSSION THREAD

**Discussion by jonathanf** [Budibase/budibase#1755] [Ideas]
Hi everyone, I think could be awesome to have in Budibase a field type or a property in text fields that allow users to provide their geolocation or select one and see them in to the apps.

This also could have a text field for search location via address via API (osm or/and Google geolocation), something like this:

![Captura de pantalla de 2021-06-19 18-01-15](https://user-images.githubusercontent.com/926635/122657350-706caf00-d128-11eb-96be-911ea3b66e67.png)

The widget could save only a field with latitude, longitude and altitude or create separated fields for each one.

Best,
Jonathan

**Comment by shogunpurple** [comment]
Hi @jonathanf, this is great feedback. We have received quite a few requests for integrations with google maps/mapbox for example.

I think a geopoint/numerical format is a nice way to represent this data, and you would be able to feed it into a map component to show different points on the map.

Just out of interest, for your use case - do you already have a list of locations available that you want to query, or is it important to be able to select and input new locations from your app? Ie. Entering an address, have it geocoded (by google or similar), then added to our database and displayed on the map?

Just trying to understand some of the map workflows our users have so we can deliver a nice experience for it.

Thanks!

**Reply by jonathanf** [reply]
Hi @shogunpurple, thanks for your quick reply! 

Yes I have the data, but this looks like 2 issues:

1. Data entry in forms (as a field/widget)
2. Report in map view maybe with mapbox widgets (as a report or view)

Here you have a data set with points, the data entry on ODK Collect/Enketo Express widget create in background this fields:

Field | Type | Use
---| --- | ---
gps_string | string | All the data
latitude | float | The lattude
longitude | float |  The longitude
altitude | float | The altitude or zero 
precision | float | The precision or zero
  
This datasets example was anonymized and the information was gather using ODK Collect (mobile) and Enketo Express (web form):

[budibase_geopoint_example.csv](https://github.com/Budibase/budibase/files/6690529/budibase_geopoint_example.csv)

Hope it helps.

Thanks!

**Reply by jonathanf** [reply]
@shogunpurple let me know if there is anything todo here

Best.

**Reply by shogunpurple** [reply]
Hi @jonathanf - sorry I missed your reply on this one. Still no development on the map component as of yet as we have been focusing on other features, but what I will do is create an issue from this discussion, so it can be tracked in our board. You can then follow along with it's progress.

**Comment by adamjehu** [comment]
Hi! @shogunpurple on the topic of geolocation, even without having google maps or mapbox integrated - perhaps an even simpler feature can be having a field which can be bound to a value like {{Device.Location}} which can pull the current lat/long from the device's location services. If something like this can be made possible, then the location can be saved to a database - allowing the coordinates to be mapped using GIS or a mapping API. Would it be possible to have a feature like this included in upcoming releases?

**Reply by mjashanks** [reply]
Hey @adamjehu - if you create a new discussion for this, and mark it as an "Idea", then we can keep track of it. It probably won't happen in the next few releases.

**Reply by adamjehu** [reply]
@mjashanks thank you, I will create it.

**Comment by tobehn** [comment]
I would also link to this idea thread #2739, just to keep it bundled.

**Comment by tobehn** [comment]
Perhaps this open data standard can be helpful, too
https://www.wikiwand.com/en/GeoJSON
https://geojson.io/

**Comment by tluyben** [comment]
Our case here would be plotting either lat lng from a table on a map or point out a bunch of fields from the table, forming a unique address,  which, fed to Google maps, would deliver coordinates drawn in the map.

Bonus for indicating which fields should appear in the mouse over, allowing a view button to view the detail and more bonus for customizing the pins and message with urls and js.

**Comment by gvlx** [comment]
Idea: 
- field is of type "Long Form Text"  and has geojson data (simpler than #2739)
- a new Design component works on this data
- new formula helpers will be needed (area, centroid, etc).

When using a database with GIS capabilities, this might not be enough.

**Comment by JoshFialkoff** [comment]
Are there updates on this? Here are a great set of fields from Gravity Forms: https://docs.gravityforms.com/form-fields/

**Reply by andz-bb** [reply]
hey @JoshFialkoff, I wrote a simple Budibase component plugin that allows users to provide their current location using the browser's Geolocation API. Once the user gives permission to show their location it sets their current coordinates to latitude and longitude form fields. You can integrate it with other components like the Embedded Map to display the coordinate data.

[Here's a link to the plugin repo](https://github.com/andz-bb/budibase-component-geolocation) and [here's how to add the plugin to your Budibase instance.](https://docs.budibase.com/docs/custom-plugin#importing-plugins) Hope this helps, happy to answer any questions you have on it!
